opnode/l2/sync_start.go

```diff
@@ -0,0 +1,111 @@
+package l2
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+
+	"github.com/ethereum-optimism/optimistic-specs/opnode/eth"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+var WrongChainErr = errors.New("wrong chain")
+
+func FindSyncStart(ctx context.Context, reference SyncReference, genesis *Genesis) (nextRefL1 eth.BlockID, refL2 eth.BlockID, err error) {
+	// 10 seconds for the whole thing (TODO: or do we want individual timeouts?)
+	ctx, cancel := context.WithTimeout(ctx, time.Second*10)
+	defer cancel()
+	var refL1 eth.BlockID
+	var parentL2 common.Hash
+	// Start at L2 head
+	refL1, refL2, parentL2, err = reference.RefByL2Num(ctx, nil, genesis)
+	if err != nil {
+		err = fmt.Errorf("failed to fetch L2 head: %v", err)
+		return
+	}
+	// Check if L1 source has the block
+	var currentL1 eth.BlockID
+	currentL1, _, err = reference.RefByL1Num(ctx, refL1.Number)
+	if err != nil {
+		if !errors.Is(err, ethereum.NotFound) {
+			err = fmt.Errorf("failed to lookup block %d in L1: %w", refL1.Number, err)
+			return
+		}
+		// if the L1 did not find the block, it might be out of sync.
+		// We cannot sync from L1 in this case, but must make sure we are not in
+		// a reorg to a L1 chain with less blocks.
+		err = nil
+		currentL1 = eth.BlockID{} // empty = not found
+	}
+	if currentL1 == refL1 {
+		// L1 node has head-block of execution-engine, so we should fetch the L1 block that builds on top.
+		var ontoL1 eth.BlockID
+		nextRefL1, ontoL1, err = reference.RefByL1Num(ctx, refL1.Number+1)
+		if err != nil {
+			// If refL1 is the head block, then we might not have a next block to build on the head
+			if errors.Is(err, ethereum.NotFound) {
+				// return the same as the engine head was already built on, no error.
+				nextRefL1 = refL1
+				refL2 = eth.BlockID{Hash: parentL2, Number: refL2.Number}
+				if refL2.Number > 0 {
+					refL2.Number -= 1
+				}
+				err = nil
+				return
```

Comment:
Okay, I get it - "the block we need" =  "the block we need to take the next sync step".

If that block doesn't exist yet and there is therefore no next sync step, then we return the parent of `refL2` and `refL1` and the logic in `driver.go / Drive` will just abort sync because `refL1 == currentL1`.

Btw, in `drive.go / Drive` we should maybe change the line `refL1, refL2, err := FindSyncStart(e.Ctx, e.SyncRef, &e.Genesis)` to be `nextRefL1, refL2` to match the variables here? (i.e. `refL1` in that line is NOT the L1 block that yielded the L2 block `refL1`).

Reply:
Yep, will fix to make the naming consistent, should be `nextRefL1` in the driver indeed.

Reply:
Actually, we need both, since the head-update needs the L1 that created the L2 block.

Reply:
Edit: we don't need both, since we combine nextRefL1 with the new derived L2 id before updating the head.